event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -41,30 +51,122 @@
 	// TODO: Change key to ObjectReference or similar
 	private final Map<String, XFunction> functions = new HashMap<>();
 
-	@Autowired // TODO: merge in here?
-	private JobLauncher launcher;
+	@Autowired
+	private FunctionDeployer deployer;
 
-	private final KubernetesClient kubernetesClient;
+	private volatile long scalingFrequencyMs = 0L; // Wait forever
 
-	public EventDispatchingHandler(KubernetesClient kubernetesClient) {
-		this.kubernetesClient = kubernetesClient;
-	}
+	private final Thread scalingThread = new Thread(new ScalingMonitor());
+
+	private final Object mutex = new Object();
+
+	private volatile boolean running = false;
+
+	private final KafkaConsumerMonitor kafkaConsumerMonitor = new KafkaConsumerMonitor(System.getenv("SPRING_CLOUD_STREAM_KAFKA_BINDER_BROKERS"));
+
+	private final SpelExpressionParser spelExpressionParser = new SpelExpressionParser();
 
 	@EventListener
-	public void onFunctionRegistered(ResourceAddedEvent<XFunction> event) {
+	public void onFunctionRegistered(ResourceAddedOrModifiedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
 		this.functions.put(functionName, functionResource);
-		this.launcher.dispatch(functionResource);
+		this.kafkaConsumerMonitor.beginTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		if (!running) {
+			running = true;
+			scalingThread.start();
+		}
+		synchronized (mutex) {
+			this.mutex.notify();
+		}
+
 		logger.info("function added: " + functionName);
 	}
 
+
 	@EventListener
 	public void onFunctionUnregistered(ResourceDeletedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
-		this.kubernetesClient.extensions().jobs().withLabel("function", functionName).delete();
```

Comment:
shouldn't we be deleting the `Deployment` in this case?

Reply:
Yes, forgot to port that part